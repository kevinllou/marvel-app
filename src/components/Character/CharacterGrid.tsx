/* eslint-disable react/no-unescaped-entities */
import { useSearchParams } from 'react-router-dom';
import CharacterCard from './CharacterCard';
import Spinner from '../Spinner/Spinner';
import useFetch from '../../hooks/useFetch';
import IApiResponse from '../../interfaces/IApiResponse';
import ICharacters from '../../interfaces/ICharacters';
import createURLFetch from '../../helpers/createURLFetch';
import IURLParams from '../../interfaces/IURLParams';
import Pagination from '../Pagination/Pagination';
import usePagination from '../../hooks/usePagination';

function CharacterGrid() {
  const [searchParams] = useSearchParams();
  const charactersFecth: IURLParams = {
    nameStartsWith: searchParams.get('nameStartsWith') || '',
    comicParam: searchParams.get('comics') || '',
    storyParam: searchParams.get('stories') || '',
    page: searchParams.get('page') || '1',
  };
  const { state, data, error } = useFetch<IApiResponse<ICharacters>>(createURLFetch(charactersFecth, 'characters'));
  const paginationRange = usePagination(
    {
      totalCount: data?.data?.total || 0,
      pageSize: 20,
      currentPage: Number(charactersFecth.page),
    },
  );
  if (state === 'loading') return <Spinner />;
  if (error) return <p>There was an error</p>;
  return (
    <>
      <div className="characters__grid">
        {data?.data?.results?.length === 0 && <p style={{ paddingTop: '.5rem' }}>We couldn't find any data ...</p>}
        {data?.data?.results?.map(({ id, name, thumbnail }) => (
          <CharacterCard key={id} name={name} thumbnail={thumbnail} id={id} />
        ))}
      </div>
      {data && paginationRange?.length > 1
      && <Pagination currentPage={charactersFecth.page} paginationRange={paginationRange || []} />}

    </>
  );
}
export default CharacterGrid;
