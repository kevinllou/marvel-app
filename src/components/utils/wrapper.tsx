import { render } from '@testing-library/react';
import { ReactElement } from 'react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import MarvelRoutes from '../../routes/MarvelRoutes';

const renderWithBrowserRouter = (component: ReactElement) => render(
  <Provider store={store}>
    <BrowserRouter>{component}</BrowserRouter>
  </Provider>,
);

const renderWithMemoryRouter = (path: string[] = ['/']) => render(
  <Provider store={store}>
    <MemoryRouter initialEntries={path}>
      <MarvelRoutes />
    </MemoryRouter>
  </Provider>,
);

export {
  renderWithBrowserRouter,
  renderWithMemoryRouter,
};
