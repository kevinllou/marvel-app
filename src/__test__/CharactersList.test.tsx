import { waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CharacterGrid from '../components/Character/CharacterGrid';
import { renderWithBrowserRouter, renderWithMemoryRouter } from '../components/utils/wrapper';
import ROUTES from '../routes/routes';

describe('Characters List Component', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  it('should render list of characters', async () => {
    const { container, getByText } = renderWithBrowserRouter(<CharacterGrid />);
    await waitFor(() => {
      expect(container.querySelector('.spinner')).not.toBeTruthy();
      expect(getByText('3-D Man')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  it('should search for hulk in the search bar', async () => {
    renderWithMemoryRouter([`${ROUTES.CHARACTERS}`]);
    await screen.findByText(/3-D Man/i);
    const searchBar = screen.getByPlaceholderText(/Search/i);
    userEvent.type(searchBar, 'Hulk{enter}');
    userEvent.click(screen.getByText('Filter'));

    await waitFor(
      () => {
        expect(screen.queryByText(/Hulk/i)).toBeInTheDocument();
        expect(screen.queryByText(/3-D Man/i)).toBeNull();
      },
    );

    userEvent.type(searchBar, '{backspace}{backspace}{backspace}{backspace}{enter}');
    userEvent.click(screen.getByText('Filter'));

    await waitFor(
      () => {
        expect(screen.queryByText(/Hulk/i)).toBeNull();
        expect(screen.queryByText(/3-D Man/i)).toBeInTheDocument();
      },
    );
  });

/*   it('should redirect to characters details', async () => {
    const { container } = renderWithMemoryRouter(['/characters']);
    const card = await screen.findByText(/3-D Man/i);
    userEvent.click(card);
    await waitFor(() => {
      expect(screen.queryByText('3-D Man')).toBeInTheDocument();
      expect(screen.queryByText('Like')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  }); */
});
