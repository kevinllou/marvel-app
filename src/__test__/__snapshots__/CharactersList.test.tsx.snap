// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters List Component should render list of characters 1`] = `
<div>
  <div
    class="characters__grid"
  >
    <div
      class="characters__gridItems"
    >
      <div
        class="characters__card"
      >
        <div
          class="characters__cardImage"
        >
          <img
            alt="3-D Man"
            src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
          />
        </div>
        <div
          class="characters__cardBody"
        >
          <div
            class="characters__name"
          >
            <a
              href="/characters/1011334"
            >
              3-D Man
            </a>
          </div>
          <div
            class="characters__options"
          >
            <button
              style="color: white;"
              type="button"
            >
              <i
                class="fa-sharp fa-solid fa-heart"
              />
            </button>
            <button
              style="color: white; pointer-events: auto;"
              type="button"
            >
              <i
                class="fa-solid fa-eye-slash"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="characters__gridItems"
    >
      <div
        class="characters__card"
      >
        <div
          class="characters__cardImage"
        >
          <img
            alt="A-Bomb (HAS)"
            src="http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16.jpg"
          />
        </div>
        <div
          class="characters__cardBody"
        >
          <div
            class="characters__name"
          >
            <a
              href="/characters/1017100"
            >
              A-Bomb (HAS)
            </a>
          </div>
          <div
            class="characters__options"
          >
            <button
              style="color: white;"
              type="button"
            >
              <i
                class="fa-sharp fa-solid fa-heart"
              />
            </button>
            <button
              style="color: white; pointer-events: auto;"
              type="button"
            >
              <i
                class="fa-solid fa-eye-slash"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <section
    class="pagination"
  >
    <button
      class="pagination__prev"
      disabled=""
      type="button"
    >
      <i
        class="fa-solid fa-angle-left"
      />
    </button>
    <button
      class="pagination__btn pagination__btn--active"
      style="pointer-events: auto;"
      type="button"
    >
      1
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      2
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      3
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      4
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      5
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: none;"
      type="button"
    >
      ...
    </button>
    <button
      class="pagination__btn"
      style="pointer-events: auto;"
      type="button"
    >
      78
    </button>
    <button
      class="pagination__next"
      type="button"
    >
      <i
        class="fa-solid fa-angle-right"
      />
    </button>
  </section>
</div>
`;
